datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -1089,6 +1090,24 @@ private void hardDeleteDatastream(Datastream ds, List<Datastream> activeStreams)
     _adapter.deleteDatastream(ds.getName());
   }
 
+  /**
+   * Invokes post datastream state change action of connector for given datastream.
+   * @param datastream the datastream
+   * @throws DatastreamException if fails to perform post datastream action
+   */
+  public void invokePostDataStreamStateChangeAction(final Datastream datastream) throws DatastreamException {
+    _log.info("Invoke post datastream state change action");
+    try {
+      final String connectorName = datastream.getConnectorName();
+      final ConnectorInfo connectorInfo = _connectors.get(connectorName);
+      connectorInfo.getConnector().postDatastreamStateChangeAction(datastream);
```

Comment:
My thinking process was same when I started implementing this logic to have the async behavior here in coordinator so that connectors do not need to worry about it. However, if you think and compare it with other existing methods from connector interface, ideally it would be each connectors responsibility to implement the way it wants. For our use case we are ok with fire and forget call, what if some one else needs a blocking call, similar to validateUpdate or postDatastreamInitialization? This made be push the async logic inside the connector and not in the upper level(coordinator). What do you think?

Reply:
Although I understand that it would be difficult to manage the life cycle of the thread from the coordinator, I was only concerned as this new method is being called while handling a client request, and I think we shouldn't ideally be supporting a blocking call outside of this repo for that.

Can we somehow enforce a max timeout for these invocations in this implementation, so that we don't block the corresponding request due to unknown implementations outside of this repo?

Reply:
You are right, this would introduce yet extra operation during REST call, however, I don't think metadata publish usecase should determine the blocking vs non-blocking call for open source use case.

> Can we somehow enforce a max timeout for these invocations in this implementation, so that we don't block the corresponding request due to unknown implementations outside of this repo?

I don't think there would be a way unless we use a separate thread and make it a blocking call with timeout(1 or 2sec). Right now, the client request timeout will kick it, which I am not sure what it is (Ideally 5sec or default 30secs). I still feel that would be inside connectors as they will be interested in the result of this call and not the coordinator.  The new method java doc explicitly calls out the behavior of this method call and if you are worried about API call latency, make sure it gets addressed in connector implementation.

I don't think it would be difficult to manage the life cycle of thread. With this design I see this as yet another helper method for connectors to do their work. If they want it to be blocking call, they can do so. For us, right now we are talking about async call for metadata publish, but that would not be a case for other connectors or other use case. For metadata publish, we can make an async call in the respective connectors, if additional functionality is needed in the future which requires blocking call, this approach will let us do that.

I am not strongly opposing this idea, but design wise it make more sense to have this within connector.

Reply:
Synced up with @hshukla on this, as long as we somehow enforce that these are implemented asynchronously, within their connector implementation, we should be good. 

And in some case when some implementations are made blocking as such that they timeout the request, it won't cause much of any harm since the state change would have been performed as expected unless in cases when multiple rest calls are invoked (in a combined fashion, which is rare).